GameServerLib/Content/Navigation/NavigationGrid.cs

```diff
@@ -758,71 +763,77 @@ public float GetHeightAtLocation(float x, float y)
         /// <returns>True = Reached destination with destination. False = Failed, with stopping position.</returns>
         public KeyValuePair<bool, Vector2> CastRay(Vector2 origin, Vector2 destination, bool checkWalkable = false, bool checkVisible = false)
         {
-            float x1 = origin.X;
-            float y1 = origin.Y;
-            float x2 = destination.X;
-            float y2 = destination.Y;
-
             // Out of bounds
-            if (x1 < MinGridPosition.X || y1 < MinGridPosition.Z || x1 >= MaxGridPosition.X || y1 >= MaxGridPosition.Z)
+            if (origin.X < MinGridPosition.X || origin.X >= MaxGridPosition.X || origin.Y < MinGridPosition.Z || origin.Y >= MaxGridPosition.Z)
             {
                 return new KeyValuePair<bool, Vector2>(false, new Vector2(float.NaN, float.NaN));
             }
 
-            float distx = x2 - x1;
-            float disty = y2 - y1;
-            float greatestdist = Math.Abs(distx);
-            if (Math.Abs(disty) > greatestdist)
-            {
-                greatestdist = Math.Abs(disty);
-            }
+            origin = TranslateToNavGrid(origin);
+            destination = TranslateToNavGrid(destination);
+
+            Vector2 dist = destination - origin;
+            float greatestdist = Math.Max(
+                Math.Abs(dist.X),
+                Math.Abs(dist.Y)
+            );
 
-            int il = (int)greatestdist;
-            float dx = distx / greatestdist;
-            float dy = disty / greatestdist;
             int i;
-            bool prevPosHadBush = HasFlag(origin, NavigationGridCellFlags.HAS_GRASS);
-            bool destinationHasGrass = HasFlag(destination, NavigationGridCellFlags.HAS_GRASS);
+            int il = (int)greatestdist;
+            Vector2 d = dist / greatestdist;
+
+            bool prevPosHadBush = HasFlag(origin, NavigationGridCellFlags.HAS_GRASS, false);
+            bool destinationHasGrass = HasFlag(destination, NavigationGridCellFlags.HAS_GRASS, false);
 
             for (i = 0; i < il; i++)
             {
+                
                 //TODO: Implement methods for maps whose NavGrids don't use SEE_THROUGH flags for buildings
-                if ((checkWalkable && !IsWalkable(x1, y1)) || (checkVisible && !IsVisible(x1, y1)))
-                {
-                    break;
-                }
-
-                bool isGrass = HasFlag(new Vector2(x1, y1), NavigationGridCellFlags.HAS_GRASS);
-                // If you are outside of a bush
-                if (checkVisible && !prevPosHadBush)
+                if (checkWalkable)
                 {
-                    if (isGrass)
+                    if(!IsWalkable(origin, translate: false))
                     {
                         break;
                     }
                 }
-
-                // If you are in a different bush
-                if (checkVisible && prevPosHadBush && destinationHasGrass)
+                else if(checkVisible)
```

Comment:
Would prefer if this option is maintained alongside `checkWalkable`.
One passthrough for both checks rather than having to call twice.

```suggestion
                if (checkVisible)
```